File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/misc/Encoder.java
package org.firstinspires.ftc.teamcode.misc;

import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.util.ElapsedTime;

/**
 * Wraps a motor instance to provide corrected velocity counts and allow reversing without changing the corresponding
 * slot's motor direction
 */
public class Encoder {
    private final static int CPS_STEP = 0x10000;

    private static double inverseOverflow(double input, double estimate) {
        double real = input;
        while (Math.abs(estimate - real) > CPS_STEP / 2.0) {
            real += Math.signum(estimate - real) * CPS_STEP;
        }
        return real;
    }

    public enum Direction {
        FORWARD(1),
        REVERSE(-1);

        private final int multiplier;

        Direction(int multiplier) {
            this.multiplier = multiplier;
        }

        public int getMultiplier() {
            return multiplier;
        }
    }

    private final DcMotorEx motor;
    private final ElapsedTime clock;

    private Direction direction;

    private int lastPosition;
    private double velocityEstimate;
    private double lastUpdateTime;

    public Encoder(DcMotorEx motor, Direction direction) {
        this.motor = motor;
        this.direction = direction;

        this.clock = new ElapsedTime();


        this.lastPosition = 0;
        this.velocityEstimate = 0.0;
        this.lastUpdateTime = clock.seconds();
    }

    public Direction getDirection() {
        return direction;
    }

    /**
     * Allows you to set the direction of the counts and velocity without modifying the motor's direction state
     *
     * @param direction either reverse or forward depending on if encoder counts should be negated
     */
    public void setDirection(Direction direction) {
        this.direction = direction;
    }

    public int getCurrentPosition() {
        int multiplier = direction.getMultiplier() * (motor.getDirection() == DcMotorSimple.Direction.FORWARD?1:-1);
        int currentPosition = motor.getCurrentPosition() * multiplier;
        if (currentPosition != lastPosition) {
            double currentTime = clock.seconds();
            double dt = currentTime - lastUpdateTime;
            velocityEstimate = (currentPosition - lastPosition) / dt;
            lastPosition = currentPosition;
            lastUpdateTime = currentTime;
        }
        return currentPosition;
    }

    public double getRawVelocity() {
        int multiplier = direction.getMultiplier() * (motor.getDirection() == DcMotorSimple.Direction.FORWARD?1:-1);
        return motor.getVelocity() * multiplier;
    }

    public double getCorrectedVelocity() {
        return inverseOverflow(getRawVelocity(), velocityEstimate);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/misc/Encoder.kt
package org.firstinspires.ftc.teamcode.misc

import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.util.ElapsedTime
import com.qualcomm.robotcore.util.MovingStatistics
import kotlin.math.abs
import kotlin.math.sign


/**
 * Wraps a motor instance to provide corrected velocity counts and allow reversing without changing the corresponding
 * slot's motor direction
 */
class Encoder(private val motor: DcMotorEx, var direction: Direction) {


    enum class Direction(val multiplier: Int) {
        FORWARD(1), REVERSE(-1);
    }
    private val CPS_STEP = 0x10000
    private fun inverseOverflow(input: Double, estimate: Double): Double {
        var real = input
        while (abs(estimate - real) > CPS_STEP / 2.0) {
            real += sign(estimate - real) * CPS_STEP
        }
        return real
    }
    private var velocityEstimates = MovingStatistics(5)

    private val clock: ElapsedTime = ElapsedTime()
    private var lastPosition: Int
    private var lastUpdateTime: Double
    val currentPosition: Int
        get() {
            val multiplier = direction.multiplier * if (motor.direction == DcMotorSimple.Direction.FORWARD) 1 else -1
            val currentPosition = motor.currentPosition * multiplier
            if (currentPosition != lastPosition) {
                val currentTime = clock.seconds()
                val dt = currentTime - lastUpdateTime
                velocityEstimates.add((currentPosition - lastPosition) / dt)
                lastPosition = currentPosition
                lastUpdateTime = currentTime
            }
            return currentPosition
        }
    val rawVelocity: Double
        get() {
            val multiplier = direction.multiplier * if (motor.direction == DcMotorSimple.Direction.FORWARD) 1 else -1
            return motor.velocity * multiplier
        }
    val correctedVelocity: Double
        get() = inverseOverflow(rawVelocity, velocityEstimates.mean)

    init {
        lastPosition = 0
        velocityEstimates = MovingStatistics(5)
        lastUpdateTime = clock.seconds()
    }
}